backend/functions/student/functions.js

```diff
@@ -36,18 +45,36 @@ const addStudentSurveyResponse = async (
   const crseIds = await Promise.all(
     courseCatalogNames.map((name) => mapCatalogNameToCrseId(name))
   );
+  var studentCrses = new Array();
+  try {
+    const ref = studentRef.doc(email);
+    studentCrses = (await ref.get()).data().groups;
+    const existingSet = new Set(studentCrses.map((crse) => crse.courseId));
+    const crsesToAdd = shouldAdd(crseIds,existingSet);
+    //creates the unioned array of courses
+    crsesToAdd.forEach((crse) => studentCrses.push({
+      courseId: crse,
+      groupNumber:-1
+    }))
+  }
+  catch(err){ 
+    crseIds.map((crseId) => {
+    studentCrses.push({
+      courseId: crseId,
+      groupNumber: -1
+    })
+   })
+  }
```

Comment:
I feel like wrapping the whole thing in a try-catch block is not the best way to approach this. We know exactly when to expect the document's data to be undefined, and can plan for it. Every new student will enter this catch block because when they submit the survey for the first time, their document doesn't exist yet.

```js
const existingData = (await studentRef.doc(email).get()).data()
const existingCourses = existingData ? existingData.groups : []
```

In this example, `existingCourses` becomes `existingData.groups` if available, otherwise `[]`.